network/network-service/server/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    logging::{LogEntry, LogSchema},
+    metrics::{increment_counter, start_timer},
+    network::NetworkServiceNetworkEvents,
+};
+use ::network::{application::storage::PeerMetadataStorage, ProtocolId};
+use aptos_config::config::NetworkServiceConfig;
+use aptos_logger::prelude::*;
+use bounded_executor::BoundedExecutor;
+use futures::stream::StreamExt;
+use network_service_types::{
+    ConnectedPeersResponse, NetworkServiceError, NetworkServiceRequest, NetworkServiceResponse,
+    Result, ServerProtocolVersionResponse,
+};
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, sync::Arc};
+use thiserror::Error;
+use tokio::runtime::Handle;
+
+mod logging;
+mod metrics;
+pub mod network;
+
+#[cfg(test)]
+mod tests;
+
+/// Network server constants
+pub const NETWORK_SERVER_VERSION: u64 = 1;
+
+#[derive(Clone, Debug, Deserialize, Error, PartialEq, Serialize)]
+pub enum Error {
+    #[error("Invalid request received: {0}")]
+    InvalidRequest(String),
+    #[error("Unexpected error encountered: {0}")]
+    UnexpectedErrorEncountered(String),
+}
+
+impl Error {
+    /// Returns a summary label for the error type
+    fn get_label(&self) -> &'static str {
+        match self {
+            Error::InvalidRequest(_) => "invalid_request",
+            Error::UnexpectedErrorEncountered(_) => "unexpected_error",
+        }
+    }
+}
+
+/// The server-side actor for the network service
+pub struct NetworkServiceServer {
+    bounded_executor: BoundedExecutor,
+    network_requests: NetworkServiceNetworkEvents,
+    peer_metadata: Arc<PeerMetadataStorage>,
+}
+
+impl NetworkServiceServer {
+    pub fn new(
+        config: NetworkServiceConfig,
+        executor: Handle,
+        network_requests: NetworkServiceNetworkEvents,
+        peer_metadata: Arc<PeerMetadataStorage>,
+    ) -> Self {
+        let bounded_executor =
+            BoundedExecutor::new(config.max_concurrent_requests as usize, executor);
+
+        Self {
+            bounded_executor,
+            network_requests,
+            peer_metadata,
+        }
+    }
+
+    /// Starts the network service server thread
+    pub async fn start(mut self) {
+        // Handle the service requests
+        while let Some(request) = self.network_requests.next().await {
+            // Log the request
+            let (peer, protocol, request, response_sender) = request;
+            debug!(LogSchema::new(LogEntry::ReceivedNetworkRequest)
+                .request(&request)
+                .message(&format!(
+                    "Received network request. Peer: {:?}, protocol: {:?}.",
+                    peer, protocol,
+                )));
+
+            // All handler methods are currently CPU-bound so we want
```

Comment:
is this really necessary?

Reply:
@zekun000, I think so. At least until we have per-application rate limiting. Otherwise, if we spawn all of these threads on a runtime used by other services (e.g., state sync, or networking), someone can trivially DOS the node by just sending a lot of requests. Also using the bounded executor means we can limit how many "concurrent" requests exist for this service.